### PR TITLE
Admin: masquerade mode: allow viewing of profile pictures for any course #8140

### LINKED ISSUE #8140: Admin: masquerade mode: allow viewing of profile pictures for any course
v5.112

image

### PR DESCRIPTION
**Pull request by tran-tien-dat**
Fixes #8140 

**Outline of Solution**

In Admin's Masquerade modes, the impersonated user is detected by the parameter user in the URL, which is not present in requests for profile pictures. Hence, currently Admin is unable to view profile pictures of team members when impersonating a student. We can solve this by either:

1. Add a user field to all the requests for profile pictures

or

2. Allow admin to have access to all profile pictures.

I choose number 2 for simplicity and also because admin should have the highest privileges.

### CODE REVIEW

File: src/main/java/teammates/logic/api/GateKeeper.java [hunk 1]
```diff
@@ -323,8 +323,12 @@ private boolean isStudent() {
         return studentsLogic.isStudentInAnyCourse(user.getNickname());
     }
 
-    public void verifyAccessibleForCurrentUserAsInstructorOrTeamMember(AccountAttributes account, String courseId,
-            String section, String email) {
+    public void verifyAccessibleForCurrentUserAsInstructorOrTeamMemberOrAdmin(AccountAttributes account, String courseId,
+                                                                              String section, String email) {
```
Comment (whipermr5) [inline src/main/java/teammates/logic/api/GateKeeper.java]: I believe we're using 8-space indentation for this

**Comment by tran-tien-dat** [comment]
According to the [Travis log](https://travis-ci.org/TEAMMATES/teammates/builds/331138340?utm_source=github_status&utm_medium=notification), the failing test at the end of the 4th try is `StudentProfilePageUiTest`. However, the test passes if I run it alone locally. Seems like the tests are interfering each other. Anyway, the changes in this PR concerns backend and should not change the UI in any way.

**Comment by tran-tien-dat** [comment]
I've managed to reproduce the failed tests, which only happens on chrome, not firefox. The test failed because when having to log in as an instructor, the UI tests uses the Admin's masquerade mode instead (by using `BaseUiTestCase.loginAdminToPage`). Hence, when the UI test expects the instructor of a different course to be unauthorised when trying to view profile pics of a student, it actually logs in as admin and then use Masquerade mode. This PR makes all profile pics available for admin, so it failed the test.

There are 2 issues here that I want to highlight:
1. Should the UI Test use masquerade mode in place of really logging in as an instructor? I think no. I think the test should emulate as closely as possible to the actual users' actions. This should be done in another PR though.
2. Should masquerade mode shows content exactly as a student see? I think yes, because that is what most people assume (such as those who wrote the UI tests). Following what most people assume will help to prevent future messing up because of false assumption. Hence, I propose to change my current approach in this PR (approach number 2) to approach number 1 (add in `user` parameter to profile pic requests)

**Comment by whipermr5** [comment]
The failure happens on Firefox too - Travis runs Firefox.

> Should the UI Test use masquerade mode in place of really logging in as an instructor? I think no. I think the test should emulate as closely as possible to the actual users' actions. This should be done in another PR though.

Correct; however I believe this can be done in this PR. I don't think it's too much work? Also it's sufficiently related to fixing the issue since it is required for the tests to pass.

> Should masquerade mode shows content exactly as a student see? I think yes, because that is what most people assume (such as those who wrote the UI tests).

Yes; however I don't think there is any page a student can access where profile pictures show up as broken due to insufficient permissions. If the student tries visiting a page via direct URL without sufficient privileges, they will get redirected to an error page, not a partially complete page with broken profile pictures.

**Comment by tran-tien-dat** [comment]
From the logs, it looks like travis runs both Firefox and Chrome? Anyway on
my computer when testing with Firefox, just keep retrying failed tests and
eventually they all pass. I also don't know why.

Regarding changing the tests, the reason why i suggest making a new PR is
because it affects other tests as well, not just these concerning student
image permissions. So it is not entirely under the purpose of this PR

For your last point, I agree with that as well, so this PR will only fixes
admin view and hopefully not screw up student view 😅.

=> so i will continue this pr by adding user=xxx to the UI and open a new
pr for fixing the tests?


On 23 Jan 2018 12:50, "John Yong" <notifications@github.com> wrote:

The failure happens on Firefox too - Travis runs Firefox.

Should the UI Test use masquerade mode in place of really logging in as an
instructor? I think no. I think the test should emulate as closely as
possible to the actual users' actions. This should be done in another PR
though.

Correct; however I believe this can be done in this PR. I don't think it's
too much work? Also it's sufficiently related to fixing the issue since it
is required for the tests to pass.

Should masquerade mode shows content exactly as a student see? I think yes,
because that is what most people assume (such as those who wrote the UI
tests).

Yes; however I don't think there is any page a student can access where
profile pictures show up as broken due to insufficient permissions. If the
student tries visiting a page via direct URL without sufficient privileges,
they will get redirected to an error page, not a partially complete page
with broken profile pictures.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/TEAMMATES/teammates/pull/8323#issuecomment-359675830>,
or mute the thread

**Comment by whipermr5** [comment]
> Regarding changing the tests, the reason why i suggest making a new PR is
> because it affects other tests as well, not just these concerning student
> image permissions.

Instead of modifying `loginAdminToPage`, make `StudentProfilePageUiTest` use `loginAsInstructor`? That way the changes can be localised to the test.

My point was to stick to the current approach. Adding an additional parameter has a higher chance of affecting the existing non-admin views, I feel.

**Comment by tran-tien-dat** [comment]
OK, now I see what you mean. However, after browsing through the code, I observe that in all tests except for those that test the login page/functionality, `loginAdminToPage` i.e. masquerade mode is always used to login as another student/instructor. For tests that has a `loginAsInstructor` function, it just uses this mechanism and calls `loginAdminToPage`. Hence, changing this behavior only here in `StudentProfilePageUiTest` does not seem like a systematic solution to me.

But well, I agree that we should minimize the changes so as not to mess things up. So I have added a `loginAsInstructorToPage` method to `StudentProfilePageUiTest`. I would be happy if that's sufficient as well, since it means less work for me 😋

**Comment by whipermr5** [comment]
> But well, I agree that we should minimize the changes so as not to mess things up. So I have added a loginAsInstructorToPage method to StudentProfilePageUiTest. I would be happy if that's sufficient as well, since it means less work for me 😋

Yup, I think it's sufficient for this PR. If you want you could open a new issue for the point you raised, though I'm not sure if the benefits are worth the effort (can be discussed in the issue thread).

File: src/test/java/teammates/test/cases/action/StudentProfilePictureActionTest.java [hunk 2]
```diff
@@ -275,34 +275,6 @@ private void testActionWithEmailAndCourseUnauthorisedInstructorOrStudentMasquera
                     + " idOfInstructor1OfCourse1 without admin permission.", uae.getMessage());
         }
 
-        gaeSimulation.loginAsAdmin("admin.user");
-        ______TS("Failure case: masqueraded instructor not from same course");
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/action/StudentProfilePictureActionTest.java]: I believe these cases should be added back? Just that they should use actual instructor accounts instead of an admin account to test.

Comment (tran-tien-dat) [reply src/test/java/teammates/test/cases/action/StudentProfilePictureActionTest.java]: There are already tests above to check for that. I think the point of these tests are exactly the opposite of our planned solutions: allowing admin to access all pictures.

Comment (whipermr5) [reply src/test/java/teammates/test/cases/action/StudentProfilePictureActionTest.java]: I think the tests are for testing what happens when instructors masquerade as students or other instructors, and not for testing admin accounts?

Comment (tran-tien-dat) [reply src/test/java/teammates/test/cases/action/StudentProfilePictureActionTest.java]: Yep. But this PR allows admin to view all pictures regardless of whether masquerade mode is on or not, so it will fail these tests.

Comment (tran-tien-dat) [reply src/test/java/teammates/test/cases/action/StudentProfilePictureActionTest.java]: Sorry for my hasty reply, which I have deleted. I don't think these tests are for that, as the 2 tests directly above are for instructor masqueraded as instructor, and student masqueraded as student. In these 2 tests actual instructor/student accounts are used. So I think if the test writer wants to use account instructor/student account, he/she could have done so?

Comment (whipermr5) [reply src/test/java/teammates/test/cases/action/StudentProfilePictureActionTest.java]: I believe the admin account was used for convenience.

Comment (tran-tien-dat) [reply src/test/java/teammates/test/cases/action/StudentProfilePictureActionTest.java]: I probably did not make my point very clear. Let me try again:

In these deleted test cases, the admin is masqueraded as either an instructor or a student. We are currently disagreeing on the purpose of these tests. From what I understand, you think that they test what happen when an instructor/student masquerades as another instructor/student. However, the 2 tests above these deleted tests have already tested for that, so these tests cannot test for the same thing?

Another way of looking at it: these tests are all testing masquerade mode, as seen in the name of the function `testActionWithEmailAndCourseUnauthorisedInstructorOrStudentMasquerade` as opposed to another function `testActionWithEmailAndCourseUnauthorisedInstructorOrStudent` also in this class. So for masquerade mode, you need: A masqueraded as B. Both A and B are clearly specified in the 2 tests above these deleted tests. However, in these deleted tests, only B is specified. For example, this line that we are commenting on is testing when `masqueraded instructor not from same course`. So who masqueraded as an instructor not from the same course? A student, another instructor, or admin? From the information written in the test, the only answer that seems reasonable is admin. Hence, I think these tests are explicitly testing what happen when *admin* masquerade as someone else, and not just using the admin masquerade mode for convenience like in the other file `StudentProfilePageUiTest`.

Sorry for the long comment. I hope I have made it clear this time.

Comment (whipermr5) [reply src/test/java/teammates/test/cases/action/StudentProfilePictureActionTest.java]: I see; my bad! 😅 Thanks for making it clear for me 👍

File: src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java [hunk 3]
```diff
@@ -297,6 +297,23 @@ private void testAjaxPictureUrl() {
         return loginAdminToPage(profileUrl, typeOfPage);
     }
 
+    private <T extends AppPage> T loginAsInstructorToPage(String instructorId, AppUrl url, Class<T> typeOfPage) {
+        //logout
+        logout();
+
+        //login
+        getHomePage().clickInstructorLogin();
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java]: Can we navigate to the `url` directly instead? This will avoid sending an additional request for the instructor home page.

File: src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java [hunk 4]
```diff
@@ -297,6 +297,23 @@ private void testAjaxPictureUrl() {
...
+
+        //login based on the login page type
+        LoginPage loginPage = AppPage.createCorrectLoginPageType(browser);
+        loginPage.loginAsInstructor(instructorId, TestProperties.TEST_INSTRUCTOR_PASSWORD);
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java]: I understand that using the instructor home page was to avoid an `InvocationTargetException` at this step, as `LoginPage#loginAsInstructor` expects an instructor home page. However, I think the right way to do things would be to overload `loginAsInstructor` with a method that accepts any page type, just like the existing `LoginPage#loginAsStudent`.

File: src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java [hunk 5]
```diff
@@ -297,6 +297,23 @@ private void testAjaxPictureUrl() {
         return loginAdminToPage(profileUrl, typeOfPage);
     }
 
+    private <T extends AppPage> T loginAsInstructorToPage(String instructorId, AppUrl url, Class<T> typeOfPage) {
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java]: This method should be moved up to `LoginPage` so it can be reused by other tests in future.

**Comment by whipermr5** [comment]
Also, note the PR naming convention (should be identical to the issue title with number) - I've helped you rename it this time.

File: src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java (hunk 5 above)
Comment (tran-tien-dat) [reply src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java]: You mean move to `BaseActionTest` right? This function is only used in testing

Comment (tran-tien-dat) [reply src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java]: I meant `BaseUiTestCase`. Sorry I was looking at the wrong file

File: src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java (hunk 4 above)
Comment (tran-tien-dat) [reply src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java]: Isn't it out of the scope of this PR, changing the behaviour of the `LoginPage`?

File: src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java (hunk 3 above)
Comment (tran-tien-dat) [reply src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java]: OK

File: src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java (hunk 5 above)
Comment (tran-tien-dat) [reply src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java]: I see. When writing this function, I was imitating the `BaseUiTestCase#loginAdminToPage` which this function replaces. I think, this function should be moved there, because `LoginPage` seems to only lower level functions like `loginAsInstructor' (fill in email address, click login) whereas this function also logout, navigate to new URL, then login.

File: src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java (hunk 4 above)
Comment (tran-tien-dat) [reply src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java]: Also, `LoginPage#loginAsStudent` also expects `StudentHomePage`?

Comment (tran-tien-dat) [reply src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java]: Actually, I think there will not be any exception, since the return value is never used. Let me try and see how it goes.

File: src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java (hunk 5 above)
Comment (whipermr5) [reply src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java]: Oh yes; I did mean `BaseUiTestCase`; my bad 😅

File: src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java (hunk 4 above)
Comment (whipermr5) [reply src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java]: There's another method: https://github.com/TEAMMATES/teammates/blob/19a0e8ee0e7d6a9e609de972257060fdee2d15f9/src/test/java/teammates/test/pageobjects/LoginPage.java#L15

File: src/test/java/teammates/test/cases/browsertests/BaseUiTestCase.java [hunk 6]
```diff
@@ -129,6 +129,25 @@ protected static Url createLocalUrl(String testFileName) throws IOException {
         return AppPage.getNewPageInstance(browser, typeOfPage);
     }
 
+    /**
+     * Logs in a page as an instructor.
+     */
+    protected <T extends AppPage> T loginAsInstructorToPage(String instructorId, AppUrl url, Class<T> typeOfPage) {
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/browsertests/BaseUiTestCase.java]: The name `loginInstructorToPage` would be more consistent with the other methods.

**Comment by tran-tien-dat** [comment]
PR updated to the latest comments.

File: src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java [hunk 7]
```diff
@@ -284,10 +284,9 @@ private void testAjaxPictureUrl() {
     private <T extends AppPage> T getProfilePicturePage(String instructorId, String email, String courseId,
                                                         Class<T> typeOfPage) {
         AppUrl profileUrl = createUrl(Const.ActionURIs.STUDENT_PROFILE_PICTURE)
-                                   .withUserId(testData.accounts.get(instructorId).googleId)
                                    .withParam(Const.ParamsNames.STUDENT_EMAIL, email)
                                    .withParam(Const.ParamsNames.COURSE_ID, courseId);
-        return loginAdminToPage(profileUrl, typeOfPage);
+        return loginInstructorToPage(instructorId, profileUrl, typeOfPage);
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java]: @tran-tien-dat this is the root cause of #8393. With this, in staging server you need a legitimate Google ID (a few, in fact) in order for this method call to succeed. If logged in as admin like previously done, masquerade mode is activated and thus a single admin Google ID suffices.

cc @whipermr5

Comment (tran-tien-dat) [reply src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java]: I think I can fix this by using the instructor account defined in `TestProperties`:

https://github.com/TEAMMATES/teammates/blob/4f57a6e62cc1e73000daecb05274b714af5c990b/src/test/java/teammates/test/driver/TestProperties.java#L32-L36

Will this solve the problem?

Comment (wkurniawan07) [reply src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java]: In principle it will, but will also create new problems since there are at least 2 different instructors logging in in this test case, and adding a new test instructor has quite a widespread effect.
To make life easier, the best course of action will be to revert back to masquerade first, and then going down that path if you really need to.

Comment (tran-tien-dat) [reply src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java]: Hmm. Changing from masquerade mode to actual instructor login is kind of necessary in this PR's approach. I think I can change the logic in the test so that it tests for the same thing but only require 1 instructor account to be logged in. (i.e. create another course for instructor 2, and then instead of instructor 2 trying to view pics from instructor 1's course, test instructor 1 trying to view pics from instructor 2's course).

Comment (wkurniawan07) [reply src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java]: Ok, that sounds like it's going to work.

Comment (whipermr5) [reply src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java]: Apologies for not catching this 😅

Comment (wkurniawan07) [reply src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java]: @whipermr5 @tran-tien-dat don't worry, I'd do this kind of mistake myself.
